dynamic_programming/subset_sum.cpp

```diff
@@ -0,0 +1,194 @@
+ * @param set 	: an array of numbers in set of size<=n
+ * @param n 	: size of the set array
+ * @param sum 	: subset sum value
+ * @param dp initilised to -1
+ * @returns true if there is a subset of set with sum equal to given sum
+ */
+template <typename T>
+bool subset_sum(const std::vector<T> &set, const T &n, const T &sum,
+						 std::vector<std::vector<T>> &dp) {
+    if (sum == 0) {   // if sum is zero , return true
+        return true;
+    }
+    if (sum < 0) {   // if sum is negative , return true
+        return false;
+	}
+    if (n < 1) {
+        return false;
+	}
+    if(dp[n][sum] != -1) {
+        return dp[n][sum];
+	}
+
+	// include last element of set in vector v and
+	// search for sunset with sum value equals sum-set[n-1] 
+    bool inc =  subset_sum(set, n - 1, sum - set[n - 1],dp);
+
+    // exclude last element of set in vector v and
+	// search for sunset with sum value equals sum
+    bool exc =  subset_sum(set, n - 1, sum,dp);
+
+    return dp[n][sum] = inc || exc;
+}
+
+}  // namespace subset_sum_topdown
+
+}  // namespace dynamic_programming
+
+/**
+ * @brief Test implementations
+ * @returns void
+ */
+static void test() {
+    // Test 1
+    const int n1 = 6;
+	std::array<int, n1> set1 = { 3, 34, 4, 12, 5, 2 };
+	const int sum1 = 9;
+    const bool ans1 =
+        dynamic_programming::subset_sum_bottomup::subset_sum(set1,sum1);
+    const bool expected_ans1 = true;
+    assert(ans1 == expected_ans1);
+    std::cout << "subset exists with sum " << sum1 
+              << std::endl;
+
+    // Test 2
+    const int n2 = 6;
+	std::array<int, n2> set2 = { 3, 34, 4, 12, 5, 2 };
+	const int sum2 = 30;
+    const bool ans2 =
+        dynamic_programming::subset_sum_bottomup::subset_sum(set2,sum2);
+    const bool expected_ans2 = false;
+    assert(ans2 == expected_ans2);
+    std::cout << "subset doesn't exists with sum " << sum2 
+              << std::endl;
+
+    // Test 3
+    const int n3 = 6;
+	std::vector<int> set3 = { 3, 34, 4, 12, 5, 2 };
+	const int sum3 = 9;
+	const int a3 = -1;
+    std::vector<std::vector<int>> dp3(n3 + 1 , std::vector<int> (sum3+1, -1));
+    const bool ans3 =
+        dynamic_programming::subset_sum_topdown::subset_sum(set3,n3,sum3,dp3);
+    const bool expected_ans3 = true;
+    assert(ans3 == expected_ans3);
+    std::cout << "subset exists with sum " << sum3 
+              << std::endl;
+
+    // Test 4
+    const int n4 = 6;
+	std::vector<int> set4 = { 3, 34, 4, 12, 5, 2 };
+	const int sum4 = 30;
+	const int a4 = -1;
+	std::vector<std::vector<int>> dp4(n4 + 1 , std::vector<int> (sum4+1, -1));
+    const bool ans4 =
+        dynamic_programming::subset_sum_topdown::subset_sum(set4,n4,sum4,dp4);
+    const bool expected_ans4 = false;
+    assert(ans4 == expected_ans4);
+    std::cout << "subset doesn't exists with sum " << sum4
+              << std::endl;
+
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // call the test function :)
```

Comment:
Adding `return 0;` at the end is a good programming practice. 🙂
```suggestion
    test();  // run self-test implementations
    return 0;
```